File: 2_Add_Two_Numbers/Add_Two_Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode *dummy = new ListNode(0); // Created a Linked list and We will save this location
        ListNode *tail = dummy;            // We dont want to lose the head, so created a operational linked List whose head is same as Dummy
        int carry = 0;
        while (l1 != nullptr || l2 != nullptr || carry != 0)
        {                                           // Checks until its value gets 0. not when the next gets 0. also will keep running until carry is there.
            int v1 = (l1 != nullptr) ? l1->val : 0; // First Checking if that position exist. if yes then saves in v1. if no then gives 0 for that position.
            int v2 = (l2 != nullptr) ? l2->val : 0; // Same here.

            int s = v1 + v2 + carry;
            carry = s / 10;                   // Checks for the carry
            int digit = s % 10;               // Gets the last digit
            tail->next = new ListNode(digit); // Creates a new node replacing the nullptr
            // This will make the node at the head unnecessary.
            tail = tail->next; // Shifts to next node
            if (l1)
                l1 = l1->next; // Checks even if it exists cause sometimes it might end before the another one, if the next node exists then only it shifts to next.
            if (l2)
                l2 = l2->next; // Same
        }
        ListNode *result = dummy->next; // Ignores that extra node at head
        delete dummy;                   // Deletes that extra node
        return result;
    }
};

File: 0002_Add_Two_Numbers/Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        dummy = ListNode(0)  # Created a Linked list and We will save this location
        tail = dummy
        carry = 0
        while l1 is not None or l2 is not None or carry != 0:
            v1 = l1.val if l1 is not None else 0
            v2 = l2.val if l2 is not None else 0

            s = v1 + v2 + carry
            carry = s // 10
            digit = s % 10
            tail.next = ListNode(digit)
            tail = tail.next
            if l1 is not None:
                l1 = l1.next
            if l2 is not None:
                l2 = l2.next
        return dummy.next
